### ISSUE TITLE
Miscalculated used memory inside the container (cgroup v2)

### ISSUE THREAD

**Issue by irnes** [lxc/lxcfs#334]
I wanted to test how memory limitation works under cgroup2  and have realized that lxcfs miscalculates used memory (/proc/meminfo) causing that the 'free' command reports the cached data as used memory rather than buff/cache as you can see in the following CLI snippet. Once I force the drop cache memory from the host it frees the used memory inside the container that is a little bit weird.

```
HOST_CGv2 ~ # free -m
              total        used        free      shared  buff/cache   available
Mem:           1989         492        1411           0          85        1461
Swap:          2047           0        2047

HOST_CGv2 ~ # lxc-attach -n mylxc 
mylxc ~ # free -m
              total        used        free      shared  buff/cache   available
Mem:           1024           7        1016           0           0        1016
Swap:          2047           0        2047

mylxc ~ # dd if=/dev/zero of=./test bs=1M count=500 # async write
500+0 records in
500+0 records out
524288000 bytes (524 MB, 500 MiB) copied, 0.569468 s, 921 MB/s

mylxc ~ # free -m
              total        used        free      shared  buff/cache   available
Mem:           1024         510         513           0           0         513
Swap:          2047           0        2047
mylxc ~ # exit
exit

HOST_CGv2 ~ # free -m
              total        used        free      shared  buff/cache   available
Mem:           1989         498         900           0         589        1454
Swap:          2047           0        2047

HOST_CGv2 ~ # echo 1 > /proc/sys/vm/drop_caches

HOST_CGv2 ~ # lxc-attach -n mylxc 
mylxc ~ # free -m
              total        used        free      shared  buff/cache   available
Mem:           1024           6        1017           0           0        1017
Swap:          2047           0        2047

```

I believe it is caused by parsing memory.stat file which unlike in cgroup v1 does not contain the 'cache' property under cgroup v2.

**Comment by brauner** [comment]
Probably yeah, and if so there's not really something we can do about it other than e.g. show `0`. Can you re-try with current master, please and report what you observe?

**Comment by fkasumovic** [comment]
@brauner @irnes I think this can still be calculated from cgroup v2 memory.stat as

```
Cached = file - shmem

Buffers = 0   // since this was zero in lxcfs for cgroup v1 as well
//  and is probably global per kernel and irelevant per cgroup
````

Now, why I think this could work good enough?
When I check kernel 5.4.x documentation, 
and assume documentation is up to date :-)
Documentation/filesystems/proc.txt

We have following description of Buffers and Cached

**Buffers:** Relatively temporary storage for raw disk blocks
              shouldn't get tremendously large (20MB or so)

**Cached:** in-memory cache for files read from the disk (the
              pagecache).  Doesn't include SwapCached

Then when I compare it to
https://www.kernel.org/doc/Documentation/cgroup-v2.txt

**file:**
	Amount of memory used to cache filesystem data,
	including tmpfs and shared memory.

**shmem:**
	Amount of cached filesystem data that is swap-backed,
	such as tmpfs, shm segments, shared anonymous mmap()s

So basically if I subtract *shmem* form *file*, I should be left with pagecache 
and maybe dentries cache its not clear from documentation on meminfo or cgroup-v2.

Let me know yours thinking, maybe I'm on something here?

**Comment by brauner** [comment]
That could work.

**Comment by irnes** [comment]
I have fixed this. 
https://github.com/lxc/lxcfs/pull/337

**Comment by brauner** [comment]
Thanks!
